src/main/java/org/truffleruby/RubyContext.java

```diff
@@ -257,7 +267,97 @@ public void initialize() {
             sharedObjects.startSharing();
         }
 
+        if (isPreInitializing()) {
+            // Cannot save the FileDescriptor in the image, referenced by the SecureRandom instance
+            random = null;
+            // Cannot save the root Java Thread instance in the image
+            threadManager.resetMainThread();
+        } else {
+            initialized = true;
+        }
+
+        this.preInitializing = false;
+    }
+
+    /** Re-initialize parts of the RubyContext depending on the current process */
+    private void reInitialize() {
+        if (reInitialized) {
+            throw new UnsupportedOperationException("re-initialization was already performed");
+        }
+
+        this.random = new SecureRandom();
+
+        this.truffleNFIPlatform = createNativePlatform();
+        coreLibrary.initializeDefaultEncodings(truffleNFIPlatform, nativeConfiguration);
+
+        threadManager.restartMainThread(Thread.currentThread());
+        threadManager.initialize(truffleNFIPlatform, nativeConfiguration);
+
+        this.reInitialized = true;
+    }
+
+    protected boolean patch(Env newEnv) {
+        this.env = newEnv;
+
+        final Options oldOptions = this.options;
+        final Options newOptions = createOptions(newEnv);
+        if (!compatibleOptions(oldOptions, newOptions)) {
+            return false;
+        }
+        this.options = newOptions;
+
+        reInitialize();
+
+        final Object toRunAtInit = Layouts.MODULE.getFields(coreLibrary.getTruffleBootModule()).getConstant("TO_RUN_AT_INIT").getValue();
+
+        for (Object proc : ArrayOperations.toIterable((DynamicObject) toRunAtInit)) {
+            String info = RubyLanguage.fileLine(language.findSourceLocation(this, proc));
+            Launcher.printTruffleTimeMetric("before-run-delayed-initialization-" + info);
+            ProcOperations.rootCall((DynamicObject) proc);
+            Launcher.printTruffleTimeMetric("after-run-delayed-initialization-" + info);
+        }
+
         initialized = true;
+        return true;
+    }
+
+    private boolean compatibleOptions(Options oldOptions, Options newOptions) {
+        final String notReusingContext = "not reusing pre-initialized context: ";
+
+        if (!newOptions.CORE_LOAD_PATH.equals(OptionsCatalog.CORE_LOAD_PATH.getDefaultValue())) {
+            Log.LOGGER.fine(notReusingContext + "-Xcore.load_path is set: " + newOptions.CORE_LOAD_PATH);
+            return false; // Should load the specified core files
+        }
+
+        // The core library captures the value of these options (via Truffle::Boot.get_option).
+        if (newOptions.NATIVE_PLATFORM != oldOptions.NATIVE_PLATFORM) {
+            Log.LOGGER.fine(notReusingContext + "-Xplatform.native is " + newOptions.NATIVE_PLATFORM);
+            return false;
+        }
+        if (newOptions.VERBOSITY != oldOptions.VERBOSITY) {
```

Comment:
Really minor, but the spacing here is inconsistent. I'd add a newline here to separate all the blocks.

Reply:
Done.